File: plugins/plugin-instagram/typescript/src/actions/send-dm.ts
/**
 * Send DM action for Instagram
 */

import type {
  Action,
  ActionExample,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { INSTAGRAM_SERVICE_NAME } from "../constants";
import type { InstagramService } from "../service";

/**
 * Action to send a direct message on Instagram
 */
export const sendDmAction: Action = {
  name: "SEND_INSTAGRAM_DM",
  description: "Send a direct message to an Instagram user",
  similes: [
    "instagram_dm",
    "instagram_message",
    "send_instagram_message",
    "dm_instagram",
    "direct_message_instagram",
  ],
  examples: [
    [
      {
        name: "{{user1}}",
        content: {
          text: "Send a message to this Instagram thread",
        },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll send that message to the Instagram thread now.",
          action: "SEND_INSTAGRAM_DM",
        },
      },
    ],
  ] as ActionExample[][],

  async validate(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
  ): Promise<boolean> {
    // Check if this is an Instagram message with a thread ID
    const content = message.content as Record<string, unknown>;
    const source = content.source as string | undefined;
    const threadId = content.threadId as string | undefined;

    if (source !== "instagram") {
      return false;
    }

    if (!threadId) {
      return false;
    }

    // Check if Instagram service is available
    const service = runtime.getService(INSTAGRAM_SERVICE_NAME);
    if (!service) {
      return false;
    }

    return true;
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ) {
    const service = runtime.getService(
      INSTAGRAM_SERVICE_NAME,
    ) as InstagramService;
    if (!service || !service.getIsRunning()) {
      if (callback) {
        await callback({
          text: "Instagram service is not running",
        });
      }
      return { success: false, error: "Service not available" };
    }

    const content = message.content as Record<string, unknown>;
    const threadId = content.threadId as string | undefined;

    if (!threadId) {
      if (callback) {
        await callback({
          text: "No thread ID provided for Instagram DM",
        });
      }
      return { success: false, error: "Missing thread ID" };
    }

    // Get response text from state or message
    const responseText =
      ((state?.response as Record<string, unknown> | undefined)?.text as
        | string
        | undefined) ||
      (content.text as string | undefined) ||
      "";

    if (!responseText) {
      if (callback) {
        await callback({
          text: "No message text to send",
        });
      }
      return { success: false, error: "Empty message" };
    }

    try {
      const messageId = await service.sendDirectMessage(threadId, responseText);

      if (callback) {
        await callback({
          text: `Message sent to Instagram thread ${threadId}`,
          action: "SEND_INSTAGRAM_DM",
        });
      }

      return { success: true, data: { threadId, messageId } };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Unknown error";
      if (callback) {
        await callback({
          text: `Failed to send Instagram DM: ${errorMessage}`,
        });
      }
      return { success: false, error: errorMessage };
    }
  },
};

File: plugins/plugin-instagram/rust/src/actions/send_dm.rs
//! Send DM action for Instagram

use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, InstagramAction};
use crate::error::Result;

/// Action to send a direct message on Instagram
pub struct SendDmAction;

#[async_trait]
impl InstagramAction for SendDmAction {
    fn name(&self) -> &'static str {
        "SEND_INSTAGRAM_DM"
    }

    fn description(&self) -> &'static str {
        "Send a direct message to an Instagram user"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        // Check if this is an Instagram message with a thread ID
        let source = context.message.get("source").and_then(|v| v.as_str());
        let has_thread = context.thread_id.is_some();
        Ok(source == Some("instagram") && has_thread)
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let response_text = context
            .state
            .get("response")
            .and_then(|v| v.get("text"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        Ok(serde_json::json!({
            "action": self.name(),
            "thread_id": context.thread_id,
            "text": response_text
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_instagram_source() {
        let action = SendDmAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "instagram",
                "text": "Hello"
            }),
            user_id: 12345,
            thread_id: Some("thread-1".to_string()),
            media_id: None,
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_non_instagram_source() {
        let action = SendDmAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "telegram",
                "text": "Hello"
            }),
            user_id: 12345,
            thread_id: Some("thread-1".to_string()),
            media_id: None,
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_no_thread() {
        let action = SendDmAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "instagram",
                "text": "Hello"
            }),
            user_id: 12345,
            thread_id: None,
            media_id: None,
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }
}
